tokio-current-thread/src/lib.rs

```diff
@@ -322,38 +321,35 @@ impl<P: Park> CurrentThread<P> {
     where
         F: Future,
     {
-        let mut enter = tokio_executor::enter().expect("failed to start `current_thread::Runtime`");
-        self.enter(&mut enter).block_on(future)
+        let _enter = tokio_executor::enter().expect("failed to start `current_thread::Runtime`");
+        self.enter().block_on(future)
     }
 
     /// Run the executor to completion, blocking the thread until **all**
     /// spawned futures have completed.
     pub fn run(&mut self) -> Result<(), RunError> {
-        let mut enter = tokio_executor::enter().expect("failed to start `current_thread::Runtime`");
-        self.enter(&mut enter).run()
+        let _enter = tokio_executor::enter().expect("failed to start `current_thread::Runtime`");
+        self.enter().run()
     }
 
     /// Run the executor to completion, blocking the thread until all
     /// spawned futures have completed **or** `duration` time has elapsed.
     pub fn run_timeout(&mut self, duration: Duration) -> Result<(), RunTimeoutError> {
-        let mut enter = tokio_executor::enter().expect("failed to start `current_thread::Runtime`");
-        self.enter(&mut enter).run_timeout(duration)
+        let _enter = tokio_executor::enter().expect("failed to start `current_thread::Runtime`");
+        self.enter().run_timeout(duration)
     }
 
     /// Perform a single iteration of the event loop.
     ///
     /// This function blocks the current thread even if the executor is idle.
     pub fn turn(&mut self, duration: Option<Duration>) -> Result<Turn, TurnError> {
-        let mut enter = tokio_executor::enter().expect("failed to start `current_thread::Runtime`");
-        self.enter(&mut enter).turn(duration)
+        let _enter = tokio_executor::enter().expect("failed to start `current_thread::Runtime`");
+        self.enter().turn(duration)
     }
 
     /// Bind `CurrentThread` instance with an execution context.
-    pub fn enter<'a>(&'a mut self, enter: &'a mut Enter) -> Entered<'a, P> {
-        Entered {
-            executor: self,
-            enter,
-        }
+    pub fn enter<'a>(&'a mut self) -> Entered<'a, P> {
```

Comment:
I don't think this fn is needed anymore. At the very least, we should make it private as keeping it would allow the user to block w/o an "enter".

Reply:
It's used by `tokio/src/runtime/current_thread/runtime.rs:213` sadly

Reply:
Ah, I think we can update `runtime.rs` to stop calling it.

- remove `tokio_executor::enter()` from [here](https://github.com/tokio-rs/tokio/blob/0d99ddd4f4adfb3ee099a5eefaea529200afcbd1/tokio/src/runtime/current_thread/runtime.rs#L199).
- remove [this line](https://github.com/tokio-rs/tokio/blob/0d99ddd4f4adfb3ee099a5eefaea529200afcbd1/tokio/src/runtime/current_thread/runtime.rs#L213).
- `f(&mut executor)` ([here](https://github.com/tokio-rs/tokio/blob/0d99ddd4f4adfb3ee099a5eefaea529200afcbd1/tokio/src/runtime/current_thread/runtime.rs#L214) stays the same.
- The fn signature [here](https://github.com/tokio-rs/tokio/blob/0d99ddd4f4adfb3ee099a5eefaea529200afcbd1/tokio/src/runtime/current_thread/runtime.rs#L188) changes to `&mut current_thread::CurrentThread<Parker>`.

I think all that should work, then the fn referenced in this comment thread can be switched to priv.

Reply:
I _think_ the `tokio_executor::enter()` call needs to stay there to avoid deadlocks, but the other changes should be sufficient I think.

Reply:
@jonhoo The calls to CurrentThread will result in enter() being called, avoiding deadlocks (i believe).